File: palindrome-linked-list/palindrome-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        int count = 0;
        ListNode run1 = head;
        ListNode stopper = null;
        ListNode holder=null;
        while(run1!=null){
            count++;
            run1=run1.next;
        }
        run1=head;
        if(count==1){
            return true;
        }
        if(count%2==0){
            count/=2;
            while(count>0){
                run1=run1.next;
                count--;
            }
            stopper = run1;
            holder=reverseLinkedList(run1);
            run1=head;
            while(run1!=stopper){
                if(run1.val!=holder.val){
                    return false;
                }
                holder=holder.next;
                run1=run1.next;
            }
        }else{
            count/=2;
            while(count>-1){
                run1=run1.next;
                count--;
            }
            stopper = run1;
            holder = reverseLinkedList(run1);
            run1=head;
            while(holder!=null){
                if(holder.val!=run1.val){
                    return false;
                }
            holder=holder.next;
            run1=run1.next;
            }
            
        }
        return true;
    }
    public static ListNode reverseLinkedList(ListNode node){
        ListNode temp = node.next;
        ListNode previous = null;
        while(node!=null){
            temp=node.next;
            node.next=previous;
            previous=node;
            node=temp;
        }
        return previous;
    }
}

File: 234-palindrome-linked-list/palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        # #Approach1
        # arr = []
        # while head != None:
        #     arr.append(head.val)
        #     head = head.next
        
        # front_pointer = 0
        # back_pointer = len(arr) - 1
        # while(front_pointer <= back_pointer):
        #     if(arr[front_pointer] != arr[back_pointer]):
        #         return False
        #     front_pointer += 1
        #     back_pointer -= 1
        
        # return True

        #Approach2
        slow = head
        fast = head
        arr=[]
        while(fast != None and fast.next != None):
            arr.append(slow.val)
            slow = slow.next
            fast = fast.next.next
        if fast == None:
            while(slow != None):
                if(slow.val != arr.pop()):
                    return False
                slow = slow.next
        elif fast.next == None:
            slow = slow.next
            while(slow != None):
                if(slow.val != arr.pop()):
                    return False
                slow = slow.next
        return True
            
            
